is_x -> qc::is x

nicky-johnstone opened the pull request:
* Updated occurrences of `is_x` calls to use the ensemble `qc::is x`.
* Updated `qc::is decimal` to account for precision and scale arguments.
* Updated and added tests.
* Tests all passed except crypt.test and db.test which didn't run due to missing package `pgtcl`.

bvw commented:
@daniel-qcode looks good to me.

daniel-qcode commented on test/is.test:
```diff
@@ -2694,219 +2724,255 @@ bmZvIDEgMCBSCi9Sb290IDEzNiAwIFIKPj4Kc3RhcnR4cmVmCjExODgxMQolJUVPRgo=}
+        qc::is uri "///$%^&m.pdf"
     } -result 0
 
+    test is-period-1.0 {qc::is period ""} {
```
Tests written like:
```
test some-test {some description} {test_code} {result}
```

or 

```
test some-test {some description} -body {test_code} -result {result}
```

Don't think it will work if we mix the two by using the -result flag but not the -body flag

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
+    proc decimal {args} {
         #| Checks if the given number is a decimal.
+        qc::args $args -precision ? -scale ? -- number
         if {[string is double -strict $number]} {
```
Is Postgresql's int range for decimals consistent with the range of doubles in TCL ?

nicky-johnstone replied:
PostgreSQL numeric/decimals:
> up to 131072 digits before the decimal point; up to 16383 digits after the decimal point

So they can get very large. Perhaps better to check the count of digits before and after a decimal point rather than checking if it's a double.

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
         return [expr {[string toupper $bool] in {Y N YES NO TRUE FALSE T F 0 1}}]
     }
 
-    proc decimal {number} {
```
@bvw Should this handle things like 1.5e10 ?

bvw replied:
Anything that is valid in Postgresql should be valid here.
Our data types and literal representation should correspond as closely as possible with pg.

daniel-qcode replied:
Something like 1.5e10 will be cast to 15000000000 by postgres.

So while the string "1.5e10" is not a valid decimal it is still castable so not sure if this proc should return o or 1 in this case ?

daniel-qcode replied:
Currently "is decimal 1.5e10" will return 1.

But "is decimal 1.5e10 11 1" will cause an error

daniel-qcode replied:
If we wanted to handler then we could do something like this to convert them into strings:

```
if { [string first e $number] } {
    set number [qc::exp2string $number]
}
```

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
+                }
+
+                set left_digit_max [expr {$precision - $scale}]
+                if { $left_count > $left_digit_max } {
```
This doesn't make any difference but I prefer something like:

```
if { $left_count > $left_digit_max } {
    return 0
} elseif { $right_count > $scale } {
    return 0
} else {
    return 1
}
```

or 

```
if { $left_count > $left_digit_max } {
    return 0
} 

if { $right_count > $scale } {
    return 0
} 

return 1
```

bvw replied:
```
if { $left_count > $left_digit_max ||  $right_count > $scale } {
    return 0
} else {
    return 1
}
```

bvw replied:
`expr {$x < pow(10,$scale-$precision)}`
to check if it fits in the space ?

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
     }
 
-    proc decimal {number} {
+    proc decimal {args} {
```
Should we raise an error if developer specifies scale but not precision ?

My first impression from tcl tests and examples was but precision and scale were both optional and not dependent on each other.

daniel-qcode replied:
Sorry misread the code I'll take another look, think how it is just now might be fine

nicky-johnstone replied:
Depends on how close we want to mimic the [PostgreSQL decimal/numeric type](http://www.postgresql.org/docs/9.4/static/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL) which only allows for precision and scale or just precision.

I'd guess that we do want to follow it as closely as possible?

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
+            if { ! [info exists precision] && ! [info exists scale] } {
+                return 1
+            } elseif { ! [info exists precision] && [info exists scale] } {
+                set count [string length [lindex [split $number .] 1]]
```
Think we could use qc::intplaces here

daniel-qcode commented on tcl/is.tcl:
```diff
@@ -534,10 +266,58 @@ namespace eval qc::is {
+                    # The leading zero wasn't given.
+                    set left_count 1
+                }
+
```
```
set left_count [string length [cast_int $number]]
set right_count [qc::intplaces $number]
```

nicky-johnstone replied:
Problem I can foresee here is that [PostgreSQL allows a massive 131072 digits before the decimal point](http://www.postgresql.org/docs/9.4/static/datatype-numeric.html#DATATYPE-NUMERIC-TABLE) so casting to integer would fail if it were out of integer range.

daniel-qcode replied:
Actually that is not going to work because cast_int is range limited

daniel-qcode replied:
yeah just realized that

On 20 January 2015 at 13:02, Nicky Johnstone <notifications@github.com>
wrote:

> In tcl/is.tcl
> <https://github.com/qcode-software/qcode-tcl/pull/157#discussion-diff-23220202>
> :
>
> > +                } else {
> > +                    return -code error "Scale must be a non-negative integer."
> > +                }
> > +            } elseif { [info exists precision] } {
> > +                if { $precision <= 0 || ! [qc::is integer $precision]} {
> > +                    return -code error "Precision must be a positive integer."
> > +                }
> > +
> > +                set parts [split $number .]
> > +                set left_count [string length [lindex $parts 0]]
> > +                set right_count [string length [lindex $parts 1]]
> > +                if { $left_count == 0 } {
> > +                    # The leading zero wasn't given.
> > +                    set left_count 1
> > +                }
> > +
>
> Problem I can foresee here is that PostgreSQL allows a massive 131072
> digits before the decimal point
> <http://www.postgresql.org/docs/9.4/static/datatype-numeric.html#DATATYPE-NUMERIC-TABLE>
> so casting to integer would fail if it were out of integer range.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qcode-software/qcode-tcl/pull/157/files#r23220202>.
>



-- 
Kind Regards,
Daniel Clark
Qcode Software Limited
www.qcode.co.uk
Tel: +44(0) 1463 896 482

nicky-johnstone commented:
Updated with discussed changes.

Also modified behaviour to follow what PostgreSQL does when only precision is given (scale seems to be defaulted to 0):

```sql
blog[DEV]=# select 3.14::numeric(5);
 numeric 
---------
       3
(1 row)
```

```tcl
% qc::is decimal(5) 3.14
0
```

Is this appropriate or do we want the above to return true?
